# Introduction to Pandas and Data Wrangling

## What is Pandas?
Pandas is a powerful open-source Python library used for data analysis and manipulation. It provides two main data structures:

* Series: One-dimensional labeled array.

* DataFrame: Two-dimensional labeled table (like an Excel sheet).

Pandas makes it easier to load, manipulate, clean, and analyze data. It is especially useful when working with CSV, Excel files, SQL databases, or even JSON.

## What is Data Wrangling?
Data wrangling, also known as data munging, is the process of cleaning and transforming raw data into a usable format. This includes:

* Removing or correcting invalid data

* Filling in missing values

* Transforming data types

* Merging and reshaping datasets

In this lesson, we will learn the basics of Pandas through hands-on exploration of a real dataset: amazonbooks.csv

* We import the pandas library.

* This allows us to use its functions to read and work with datasets.

In [2]:
import pandas



---



* This command reads the amazonbooks.csv file.

* The encoding is set to ISO-8859-1 because the file contains special characters that can cause errors if UTF-8 encoding is used.

* However, since we didn’t store the output, we won’t see it or be able to use it later.

In [3]:
pandas.read_csv('amazonbooks.csv', encoding="ISO-8859-1") # "utf-8" is the default
# file doesnt load without passing `parameter=argument` `encoding="ISO-8859-1"` to .read_csv()

Title               Author  \
0    1,001 Facts that Will Scare the S#*t Out of Yo...          Cary McNeal   
1    21: Bringing Down the House - Movie Tie-In: Th...          Ben Mezrich   
2         100 Best-Loved Poems (Dover Thrift Editions)                Smith   
3              1421: The Year China Discovered America        Gavin Menzies   
4      1493: Uncovering the New World Columbus Created      Charles C. Mann   
..                                                 ...                  ...   
320                            Where the Sidewalk Ends     Shel Silverstein   
321                                    White Privilege  Paula S. Rothenberg   
322                                Why I wore lipstick        Geralyn Lucas   
323  Worlds Together, Worlds Apart: A History of th...        Robert Tignor   
324                                  Wuthering Heights         Emily Bronte   

     List Price  Amazon Price Hard_or_Paper  NumPages               Publisher  \
0         12.95          5.18             P     304.0             Adams Media   
1         15.00         10.20             P     273.0              Free Press   
2          1.50          1.50             P      96.0      Dover Publications   
3         15.99         10.87             P     672.0        Harper Perennial   
4         30.50         16.77             P     720.0                   Knopf   
..          ...           ...           ...       ...                     ...   
320       18.99         12.24             H     192.0           HarperCollins   
321       27.55         27.55             P     160.0        Worth Publishers   
322       12.95          5.18             P     224.0     St Martin's Griffin   
323       97.50         97.50             P     480.0  W. W. Norton & Company   
324       16.99          4.95             P     344.0              CreatSpace   

     Pub year     ISBN-10  Height  Width  Thick  Weight_oz  
0      2010.0  1605506249     7.8    5.5    0.8       11.2  
1      2008.0  1416564195     8.4    5.5    0.7        7.2  
2      1995.0   486285537     8.3    5.2    0.3        4.0  
3      2008.0    61564893     8.8    6.0    1.6       28.8  
4      2011.0   307265722     8.0    5.2    1.4       22.4  
..        ...         ...     ...    ...    ...        ...  
320    2004.0    60572345     9.3    6.6    1.1       24.0  
321    2011.0  1429233443     9.1    6.1    0.7        8.0  
322    2005.0  031233446X     8.0    5.4    0.7        6.4  
323    2010.0   393934942    10.7    8.9    0.9       14.4  
324    2011.0  1463533411     7.8    5.1    1.0       16.0  

[325 rows x 13 columns]



---



* This will cause an error because pd has not yet been defined.

* It assumes we used import pandas as pd, which hasn’t been done yet.

In [ ]:
pd.read_csv('amazonbooks.csv', encoding="ISO-8859-1")



---



* This is the recommended way to import Pandas.

* `pd` is a common alias that saves us from writing pandas. every time.

In [4]:
import pandas as pd # This alias is the standard way to use import/use/call pandas stuff



---



* Now it works perfectly.

* It loads and displays the dataset using the alias `pd`.

* Still, we haven’t saved it to a variable.

In [5]:
pd.read_csv('amazonbooks.csv', encoding="ISO-8859-1")

Title               Author  \
0    1,001 Facts that Will Scare the S#*t Out of Yo...          Cary McNeal   
1    21: Bringing Down the House - Movie Tie-In: Th...          Ben Mezrich   
2         100 Best-Loved Poems (Dover Thrift Editions)                Smith   
3              1421: The Year China Discovered America        Gavin Menzies   
4      1493: Uncovering the New World Columbus Created      Charles C. Mann   
..                                                 ...                  ...   
320                            Where the Sidewalk Ends     Shel Silverstein   
321                                    White Privilege  Paula S. Rothenberg   
322                                Why I wore lipstick        Geralyn Lucas   
323  Worlds Together, Worlds Apart: A History of th...        Robert Tignor   
324                                  Wuthering Heights         Emily Bronte   

     List Price  Amazon Price Hard_or_Paper  NumPages               Publisher  \
0         12.95          5.18             P     304.0             Adams Media   
1         15.00         10.20             P     273.0              Free Press   
2          1.50          1.50             P      96.0      Dover Publications   
3         15.99         10.87             P     672.0        Harper Perennial   
4         30.50         16.77             P     720.0                   Knopf   
..          ...           ...           ...       ...                     ...   
320       18.99         12.24             H     192.0           HarperCollins   
321       27.55         27.55             P     160.0        Worth Publishers   
322       12.95          5.18             P     224.0     St Martin's Griffin   
323       97.50         97.50             P     480.0  W. W. Norton & Company   
324       16.99          4.95             P     344.0              CreatSpace   

     Pub year     ISBN-10  Height  Width  Thick  Weight_oz  
0      2010.0  1605506249     7.8    5.5    0.8       11.2  
1      2008.0  1416564195     8.4    5.5    0.7        7.2  
2      1995.0   486285537     8.3    5.2    0.3        4.0  
3      2008.0    61564893     8.8    6.0    1.6       28.8  
4      2011.0   307265722     8.0    5.2    1.4       22.4  
..        ...         ...     ...    ...    ...        ...  
320    2004.0    60572345     9.3    6.6    1.1       24.0  
321    2011.0  1429233443     9.1    6.1    0.7        8.0  
322    2005.0  031233446X     8.0    5.4    0.7        6.4  
323    2010.0   393934942    10.7    8.9    0.9       14.4  
324    2011.0  1463533411     7.8    5.1    1.0       16.0  

[325 rows x 13 columns]



---



* Here we load the dataset and save it in a variable named `amazonbooks`.

* Now we can reuse this dataset throughout our analysis.

In [9]:
amazonbooks = pd.read_csv('amazonbooks.csv', encoding="ISO-8859-1")



---



* Returns `<class 'pandas.core.frame.DataFrame'>`

* This confirms that `amazonbooks` is a Pandas DataFrame.

In [10]:
type(amazonbooks)

pandas.core.frame.DataFrame



---



* This shows the type of data in each column (e.g., integer, float, object).

* Useful for understanding what kind of operations you can perform on each column.

In [11]:
amazonbooks.dtypes
# 3: int
# 3.14: float / float64

Title             object
Author            object
List Price       float64
Amazon Price     float64
Hard_or_Paper     object
NumPages         float64
Publisher         object
Pub year         float64
ISBN-10           object
Height           float64
Width            float64
Thick            float64
Weight_oz        float64
dtype: object



---



* Displays the full dataset.

* Allows you to visually inspect rows, columns, and values.

In [12]:
amazonbooks

Title               Author  \
0    1,001 Facts that Will Scare the S#*t Out of Yo...          Cary McNeal   
1    21: Bringing Down the House - Movie Tie-In: Th...          Ben Mezrich   
2         100 Best-Loved Poems (Dover Thrift Editions)                Smith   
3              1421: The Year China Discovered America        Gavin Menzies   
4      1493: Uncovering the New World Columbus Created      Charles C. Mann   
..                                                 ...                  ...   
320                            Where the Sidewalk Ends     Shel Silverstein   
321                                    White Privilege  Paula S. Rothenberg   
322                                Why I wore lipstick        Geralyn Lucas   
323  Worlds Together, Worlds Apart: A History of th...        Robert Tignor   
324                                  Wuthering Heights         Emily Bronte   

     List Price  Amazon Price Hard_or_Paper  NumPages               Publisher  \
0         12.95          5.18             P     304.0             Adams Media   
1         15.00         10.20             P     273.0              Free Press   
2          1.50          1.50             P      96.0      Dover Publications   
3         15.99         10.87             P     672.0        Harper Perennial   
4         30.50         16.77             P     720.0                   Knopf   
..          ...           ...           ...       ...                     ...   
320       18.99         12.24             H     192.0           HarperCollins   
321       27.55         27.55             P     160.0        Worth Publishers   
322       12.95          5.18             P     224.0     St Martin's Griffin   
323       97.50         97.50             P     480.0  W. W. Norton & Company   
324       16.99          4.95             P     344.0              CreatSpace   

     Pub year     ISBN-10  Height  Width  Thick  Weight_oz  
0      2010.0  1605506249     7.8    5.5    0.8       11.2  
1      2008.0  1416564195     8.4    5.5    0.7        7.2  
2      1995.0   486285537     8.3    5.2    0.3        4.0  
3      2008.0    61564893     8.8    6.0    1.6       28.8  
4      2011.0   307265722     8.0    5.2    1.4       22.4  
..        ...         ...     ...    ...    ...        ...  
320    2004.0    60572345     9.3    6.6    1.1       24.0  
321    2011.0  1429233443     9.1    6.1    0.7        8.0  
322    2005.0  031233446X     8.0    5.4    0.7        6.4  
323    2010.0   393934942    10.7    8.9    0.9       14.4  
324    2011.0  1463533411     7.8    5.1    1.0       16.0  

[325 rows x 13 columns]



---



* `describe()` gives us summary statistics for all numerical columns.

* We can quickly see values like:

   * count: number of non-null entries

   * mean, std (standard deviation)

   * min, max, and quartiles (25%, 50%, 75%)

* This is essential for understanding the spread and central tendency of our data.

In [14]:
amazonbooks.describe()

List Price  Amazon Price    NumPages     Pub year      Height  \
count  324.000000    325.000000  323.000000   324.000000  321.000000   
mean    18.579753     13.333846  335.857585  2002.206790    8.163240   
std     14.252829     13.727679  161.984389    10.629002    0.918739   
min      1.500000      0.770000   24.000000  1936.000000    5.100000   
25%     13.950000      8.600000  208.000000  1998.000000    7.900000   
50%     15.000000     10.200000  320.000000  2005.000000    8.100000   
75%     19.950000     13.130000  416.000000  2010.000000    8.500000   
max    139.950000    139.950000  896.000000  2011.000000   12.100000   

            Width       Thick   Weight_oz  
count  320.000000  324.000000  316.000000  
mean     5.585000    0.907716   12.487975  
std      0.874057    0.368625    6.644648  
min      4.100000    0.100000    1.200000  
25%      5.200000    0.600000    7.800000  
50%      5.400000    0.900000   11.200000  
75%      5.900000    1.100000   16.000000  
max      9.500000    2.100000   35.200000



---



* This code checks the entire DataFrame `amazonbooks` for missing (null/NaN) values.

* It returns a DataFrame of the same shape, but instead of the original data, it contains:

   * True where the value is missing (i.e., is NaN)

   * False where the value is present


* In data analysis, missing data is a common issue. For example:

   * A book might not have a listed price.

   * An author field might be empty for some entries.

* This command helps you identify which cells are missing values, so you can decide how to handle them later (e.g., drop rows, fill with default values, etc.).

In [13]:
amazonbooks.isnull()

Title  Author  List Price  Amazon Price  Hard_or_Paper  NumPages  \
0    False   False       False         False          False     False   
1    False   False       False         False          False     False   
2    False   False       False         False          False     False   
3    False   False       False         False          False     False   
4    False   False       False         False          False     False   
..     ...     ...         ...           ...            ...       ...   
320  False   False       False         False          False     False   
321  False   False       False         False          False     False   
322  False   False       False         False          False     False   
323  False   False       False         False          False     False   
324  False   False       False         False          False     False   

     Publisher  Pub year  ISBN-10  Height  Width  Thick  Weight_oz  
0        False     False    False   False  False  False      False  
1        False     False    False   False  False  False      False  
2        False     False    False   False  False  False      False  
3        False     False    False   False  False  False      False  
4        False     False    False   False  False  False      False  
..         ...       ...      ...     ...    ...    ...        ...  
320      False     False    False   False  False  False      False  
321      False     False    False   False  False  False      False  
322      False     False    False   False  False  False      False  
323      False     False    False   False  False  False      False  
324      False     False    False   False  False  False      False  

[325 rows x 13 columns]



---



This line tells us how many missing values are present in each column of the `amazonbooks` DataFrame.

* ✅ It builds on the previous command (`amazonbooks.isnull()`) which returned a DataFrame of True (missing) and False (not missing).

* 🧮 Now, it sums up the True values column-wise:

* In Python and pandas:

   * `True` is automatically treated as 1

   * `False` is automatically treated as 0

* So `.sum()` just adds the 1s in each column → which gives us the count of missing values per column.

In [15]:
# True=1 and False=0 coercion
amazonbooks.isnull().sum() # axis=0 is the default param=arg to the .sum() function

Title            0
Author           1
List Price       1
Amazon Price     0
Hard_or_Paper    0
NumPages         2
Publisher        1
Pub year         1
ISBN-10          0
Height           4
Width            5
Thick            1
Weight_oz        9
dtype: int64

✅ Boolean coercion:
Python treats `True` as `1` and `False` as `0` in mathematical operations.

✅ `.sum()` on a DataFrame:
By default, it adds values down each column (`axis=0`), which is what we want here.

If you use `axis=1`, it would add across each row (which is useful in other cases).



---



This command shows us the first few rows of the `amazonbooks` DataFrame — specifically, the first 5 rows by default.

Think of it like peeking into the top of your spreadsheet to get a feel for what kind of data you’re working with.

In [16]:
amazonbooks.head()

Title           Author  \
0  1,001 Facts that Will Scare the S#*t Out of Yo...      Cary McNeal   
1  21: Bringing Down the House - Movie Tie-In: Th...      Ben Mezrich   
2       100 Best-Loved Poems (Dover Thrift Editions)            Smith   
3            1421: The Year China Discovered America    Gavin Menzies   
4    1493: Uncovering the New World Columbus Created  Charles C. Mann   

   List Price  Amazon Price Hard_or_Paper  NumPages           Publisher  \
0       12.95          5.18             P     304.0         Adams Media   
1       15.00         10.20             P     273.0          Free Press   
2        1.50          1.50             P      96.0  Dover Publications   
3       15.99         10.87             P     672.0    Harper Perennial   
4       30.50         16.77             P     720.0               Knopf   

   Pub year     ISBN-10  Height  Width  Thick  Weight_oz  
0    2010.0  1605506249     7.8    5.5    0.8       11.2  
1    2008.0  1416564195     8.4    5.5    0.7        7.2  
2    1995.0   486285537     8.3    5.2    0.3        4.0  
3    2008.0    61564893     8.8    6.0    1.6       28.8  
4    2011.0   307265722     8.0    5.2    1.4       22.4

* `.head()` is a pandas method used to preview data.

* By default, it returns the first 5 rows, but you can customize it:

   * Example: `amazonbooks.head(10)` → shows the first 10 rows.



---



## 🧠 What is .iloc?
* `.iloc[]` stands for “integer-location based indexing”

* It lets you select rows and columns by their numerical position, not by name.

Think of it like saying: “Give me the first few rows and columns by number, not by label.”

In [ ]:
amazonbooks.iloc[:5,:] # amazonbooks.iloc[0:5,0:13]

This means:

* `:5` → rows from index 0 up to (but not including) 5, i.e., the first 5 rows

* `:` → all columns

✅ Result: The first 5 rows and all columns of the amazonbooks DataFrame.

💡 Pro Tip:
* Use `.iloc[::]` to slice with row and column numbers

* Use `.loc[::]` to slice with row and column labels



---



### How to detect rows with multiple missing values using a logical condition.

1. amazonbooks.isnull()
   * Returns a DataFrame of the same shape as amazonbooks

   * Each cell becomes:

    * `True` if the original value is missing (NaN)

    * `False` if the value exists

2. `.sum(axis=1)`
   * This sums the True values row-wise (`axis=1` means across columns)

   * Since `True = 1` and `False = 0`, this gives the number of missing values in each row

3. `> 1`
   * Compares each row’s missing value count

   * Returns:

    * `True` for rows with more than one missing value

    * `False` otherwise

✅ Final Result: A Boolean Series where:

* Each index corresponds to a row in amazonbooks

* The value is True if that row has more than one missing value

In [ ]:
amazonbooks.isnull().sum(axis=1)>1

🧠 Why is this useful?
In data wrangling, we often want to:

* Flag incomplete records

* Decide whether to drop or impute them

This line is a quick filter for finding rows that might be too incomplete to use directly.



---



🧠 Understanding the logic inside .iloc[ ... , :]

`amazonbooks.isnull().sum(axis=1) > 1`

* As explained earlier, this returns a Boolean Series

* Each `True` indicates a row with more than one missing value

🧠 Passing the Boolean Series to .iloc

`amazonbooks.iloc[ boolean_series , : ]`

* `.iloc[]` is designed for integer positions only, not labels or Boolean masks.

* So this line raises an error (that’s why it says # breaks!)

* It tries to use the Boolean Series with .iloc, but .iloc expects integer index positions, not True/False.



In [ ]:
# breaks!
amazonbooks.iloc[ amazonbooks.isnull().sum(axis=1)>1 ,:]



---



📌 Step 1: `amazonbooks.isnull()`

This returns a DataFrame of the same shape as amazonbooks, but with:

* `True` where a cell is missing (i.e., NaN)

* `False` where the data is present

👉 It’s like a "missing value detector" 🔍



📌 Step 2: `.sum(axis=1)`

This adds up the `True` values row-wise:

* In Python, `True = 1` and `False = 0`

* So this step counts the number of missing values in each row

For example, if a row has 2 missing values → this sum would be `2`.


📌 Step 3: `> 1`
This returns a Boolean Series:

* `True` for rows that have more than 1 missing value

* `False` otherwise


📌 Step 4: `amazonbooks[ ... ]`

This is where we filter the DataFrame using the Boolean Series we just built.

* Only rows with more than one missing value are displayed


✅ Result
This cell shows you exactly which rows in the `amazonbooks` dataset have multiple missing entries.

In [ ]:
amazonbooks[amazonbooks.isnull().sum(axis=1)>1]

Imagine you have a table of students and their quiz scores across 5 weeks. If a student missed more than 1 quiz (i.e., has more than 1 blank score), you want to check who they are.

This cell helps you spot those rows easily!



---



> * `0`	At least 1 missing value	1 or more NaNs
> * `1`	More than 1 missing value	2 or more NaNs

In [ ]:
amazonbooks[amazonbooks.isnull().sum(axis=1)>0]

Imagine you're looking at student assignments:

> 0 → Find all students who missed at least 1 question.

> 1 → Find all students who missed more than 1 question.



---



This code is filtering the dataset to return only the rows with absolutely no missing values.

* `amazonbooks.isnull()`	Returns a DataFrame of booleans, where each cell is `True` if it's missing (`NaN`) and `False` otherwise.

* `.sum(axis=1)`	Adds up all the `True` values per row. Since `True = 1`, this gives the count of NaNs per row.

* `== 0`	Checks where the count of missing values is 0 – i.e., no missing data in that row.

* Final result: `amazonbooks[...]`	Filters and returns only the fully complete rows – the cleanest part of your data!

In [ ]:
amazonbooks[amazonbooks.isnull().sum(axis=1)==0]

Imagine checking attendance sheets for classes. This line is like saying:

> “Show me all the students who attended all sessions (i.e., no blanks at all)!”

💡 Why Is This Useful?
* These rows are ready for immediate analysis – no missing values to fill or drop.

* You may want to analyze only this “perfect subset” of the data, especially for:

   * Initial visualizations

   * Model training

   * Descriptive stats



---



## SQL: structured query language
- `select` [columns] (`.loc[:,]` which supports "boolean indexing" unlike `.iloc[]`)
- `where` [rows] (`[]` or `.loc[:,]` but not `.iloc[]`]
- `groupby` then "aggregation function"

In [ ]:
amazonbooks.select('Author')

❌ Why It Breaks
In pandas, there is no .select() method for DataFrames.

Unlike some other tools or languages (like SQL or PySpark, which do use .select()), pandas uses:

* Square brackets for column access

* `.loc` or `.iloc` for more advanced selections



---



🎯 Goal of the Line:
> “Show me the rows in the amazonbooks DataFrame that contain any missing values, but only show the Author and Title columns.”

In [ ]:
# tuple (1,2,3) # immutable
# list [1,2,3] # mutable
# In pandas, either can be used to select multiple columns, but lists are more common.
amazonbooks.loc[amazonbooks.isnull().sum(axis=1)>0, ('Author', 'Title')]

* This uses .loc[row_filter, column_selector]

* It returns only the rows with missing data, and only the Author and Title columns





---



In [ ]:
# amazonbooks.filter(amazonbooks.isnull().sum(axis=1)>0) # WRONG -- not using this one right...
# amazonbooks.where(amazonbooks.isnull().sum(axis=1)>0) # WRONG AGAIN!!
# `.where()` is actually a cool useful function to replace specific values
# e.g., like missing values!
# ... hmm ... actually ...
# Maybe the best way to fill in missing values would be the `.fillna()` function


### 🧠 What is .groupby() in Pandas?
*Think of `.groupby()` as a way to split your dataset into smaller groups based on the values in a column (in this case, Publisher).

* Once grouped, you can apply aggregation functions like `.count()`, `.mean()`, `.sum()`, etc.

> To organize the `amazonbooks` dataset into groups where each group contains books published by the same publisher.

In [ ]:
amazonbooks.groupby('Publisher')

Imagine you’re organizing a library.
You decide to arrange books by publisher.
Now you have stacks: one for “Penguin”, another for “O’Reilly”, and so on.

`.groupby('Publisher')` is like building those stacks — but nothing is computed yet!



---



This counts the number of valid (non-null) values for each column within each group.



In [ ]:
amazonbooks.groupby('Publisher').count()



---



To count the total number of books listed for each publisher, regardless of whether any column values are missing.

* `.size()`
This counts the number of rows (books) in each group.
It includes rows with missing values.

In [ ]:
amazonbooks.groupby('Publisher').size()



---



To list publishers from the dataset in order of how many books they have, from the least to the most.



In [ ]:
amazonbooks.groupby('Publisher').size().sort_values() # default is `ascending=True`



---



To list publishers from the dataset in order of how many books they have, from the most to the least.



In [ ]:
amazonbooks.groupby('Publisher').size().sort_values(ascending=False) # False not false or FALSE



---



To find out how many publishers have a certain number of books.

> 💡 Think: How many publishers have exactly 1 book? How many have exactly 2?

In [ ]:
amazonbooks.groupby('Publisher').size().value_counts() # count_values



---



To find out how many books were published in each year.



In [ ]:
amazonbooks.groupby('Pub year').size()#.value_counts()



---



To get a statistical summary of all numerical columns in the amazonbooks DataFrame.



In [ ]:
amazonbooks.describe()

> `.describe()` automatically selects all numeric columns (like integers or floats).

> It then returns summary statistics for each numeric column.

`count`	Number of non-missing (non-NaN) values

`mean`	Average value

`std`	Standard deviation (spread/variation)

`min`	Smallest value

`25%`	25th percentile (Q1) – lower quartile

`50%` (median)	50th percentile (Q2) – middle value

`75%`	75th percentile (Q3) – upper quartile

`max`	Largest value



---



To display the entire DataFrame named `amazonbooks`.



In [ ]:
amazonbooks



---



To select a single column named "`NumPages`" from the `amazonbooks` DataFrame.

* This retrieves the `NumPages` column, which likely contains the number of pages for each book in the dataset.

* The result is a **pandas Series**, not a **DataFrame**.


In [ ]:
amazonbooks[("NumPages")]



---



In [ ]:
amazonbooks[(("NumPages"))] # tuple in a tuple () immutable



---



To extract the "`NumPages`" column as a **DataFrame** instead of a Series.



In [ ]:
amazonbooks[[("NumPages")]] # tuple in a list [] mutable



---



In [ ]:
amazonbooks.NumPages # amazonbooks[("NumPages")] / amazonbooks[(("NumPages"))]



---



To filter or create a Boolean Series that checks whether each book has more than the average number of pages.



In [ ]:
amazonbooks.NumPages > 335.857585



---



1. Boolean expressions as group keys:

 * `amazonbooks.NumPages > 335.857585` returns `True` or `False` for each book.

 * These values are used as groups.

2. `.groupby(...)` groups the DataFrame based on this Boolean condition.

3. `.size()` returns the number of rows (books) in each group.

In [ ]:
amazonbooks.groupby(amazonbooks.NumPages > 335.857585).size()

Which means:

* 📕 167 books have NumPages <= 335.857585

* 📘 158 books have NumPages > 335.857585



---



1. Grouping by a Boolean Condition:

* `amazonbooks.NumPages > 335.857585` returns a Boolean Series.

* This splits the data into two groups based on that condition.

2. `.mean()` on groupby object:

* Calculates the average of all numeric columns for each group.

* Ignores non-numeric columns like `Author`, `Title`, etc.

In [ ]:
amazonbooks.groupby(amazonbooks.NumPages > 335.857585).mean()



---



🧠 What's New Here?

✅ `numeric_only=True`:

* This parameter explicitly tells pandas to only compute the mean of numeric columns, avoiding any warning or ambiguity.

* Useful in newer versions of pandas (>=1.5) where some non-numeric columns might sneak into `.mean()` by accident.

* Ensures clean and consistent output.



In [ ]:
amazonbooks.groupby(amazonbooks.NumPages > 335.857585).mean(numeric_only=True)



---



🎯 What’s the Goal?

We want to find out:

> Is there a difference in average Amazon price between shorter books and longer books?


🔹 `amazonbooks[['Amazon Price']]`:

* This selects just the Amazon Price column as a DataFrame (not a Series).

* Why double brackets? Because we're passing a list of columns — even though it has only one item, it keeps it as a DataFrame (which is expected by `.groupby()` + `.mean()`).

🔹 .groupby(amazonbooks.NumPages > 335.857585):

* This creates two groups:

   * `False`: books with 335.857585 pages or fewer

   * `True`: books with more than 335.857585 pages

* We're essentially splitting books into “short” and “long” categories.

🔹 `.mean()`:

* For each group, it calculates the average Amazon Price.

* It tells us:

 > “Do longer books tend to cost more?”

In [ ]:
amazonbooks[['Amazon Price']].groupby(amazonbooks.NumPages > 335.857585).mean()



---



We want to create a new column in the dataset that categorizes each book as having above average page count or not.


🔹 `amazonbooks.NumPages > 335.857585`

* This is a boolean condition.

* It returns a `True` or `False` for each row, depending on whether that book has more pages than the average number of pages (in this case, ≈ 335.86).

* Example:

 * 500 pages → `True`

 * 200 pages → `False`

🔹 `amazonbooks['AboveAverageNumPages'] = ...`

* This creates a new column in the `amazonbooks` DataFrame.

* The column will store `True/False` values for each book.

* It becomes a flag or categorical label that we can use for grouping, filtering, and visualization.

In [ ]:
# add a new column!
amazonbooks['AboveAverageNumPages'] = amazonbooks.NumPages > 335.857585



---



🧠 What Are We Doing?
We are using grouping and aggregation to compare the average values (like price, number of pages, etc.) between two groups:

* Books with above-average number of pages

* Books with average or below-average number of pages

In [ ]:
amazonbooks.groupby('AboveAverageNumPages').mean(numeric_only=True)

🔸 `amazonbooks.groupby('AboveAverageNumPages')`

* This groups the dataset into two categories:

 * `False` → Books with 335 or fewer pages

 * `True` → Books with more than 335 pages

🔸 `.mean(numeric_only=True)`

* For each group (`True` and `False`), this computes the average (mean) of all numeric columns (like `Amazon Price`, `NumPages`, `Pub year`, etc.)

* `numeric_only=True` ensures that only numeric data is included in the mean calculation (e.g., ignores text like book titles or authors).



---



In [ ]:
# breaks!
amazonbooks[['Amazon Price']].groupby('AboveAverageNumPages').mean()



---



💡 What’s Happening Here?
We are:

1. Selecting two columns:

 * `'Amazon Price'`: our numeric variable of interest

 * `'AboveAverageNumPages'`: a boolean column created earlier (True if a book has more pages than average)

2. Grouping the data by `'AboveAverageNumPages'` (True or False)

3. Calculating the average for all numeric columns in each group

In [ ]:
amazonbooks[['Amazon Price','AboveAverageNumPages']].groupby('AboveAverageNumPages').mean()

🟢 `[['Amazon Price', 'AboveAverageNumPages']]`

* Selects both columns — one numeric and one boolean

* The boolean column (`AboveAverageNumPages`) is the one we'll group by

🔷 `.groupby('AboveAverageNumPages')`

* Creates two groups:

 * `False`: Books with average or below-average number of pages

 * `True`: Books with above-average number of pages

📘 `.mean()`

* Calculates the average Amazon Price for each group

* *Also tries to compute the average of `AboveAverageNumPages` itself (it will be 0.0 or 1.0)*



---



In [ ]:
amazonbooks



---



🎯 What’s the Goal Here?

👉 To calculate the average cost per page for different types of books — hardcover vs paperback — and compare them.

In [ ]:
#amazonbooks['PricePerPage'] = amazonbooks.NumPages/amazonbooks['Amazon Price'] # oops! upside down
amazonbooks['PricePerPage'] = amazonbooks['Amazon Price']/amazonbooks.NumPages
amazonbooks[['PricePerPage','Hard_or_Paper']].groupby('Hard_or_Paper').mean()



---



🧠 Quick Refresher: What is Standard Deviation?

* Standard deviation tells us how much variation there is from the average (mean).

* A low standard deviation means the values are close to the mean (less spread out).

* A high standard deviation means the values are more spread out (more variability).

In [ ]:
# The "standard deviation" `.std()` function measure how spread around the data mean the data is
amazonbooks[['PricePerPage','Hard_or_Paper']].groupby('Hard_or_Paper').std()

✅ Interpretation:

* The standard deviation is slightly higher for paperback books.

* This means paperback books have a more varied (less consistent) cost per page.

* In contrast, hardcover books have more consistent pricing per page.



---



In [ ]:
amazonbooks

### What we learned

- boolean indexing
- `.iloc[]` selecting based on index
- `.loc[]` selecting with boolean indexing or columns or both
- `[]` selecting with boolean indexing or columns but not both
- `.groupby()` and "aggregation functions" like ~~`.count()`~~ `.size()`, `.mean()`, and `.std()`
- adding/creating columns
- learned about accessing columns
    - `df.col` or `df['col']` or `df[('col')]` or `df[(('col'))]` give series
    - `df[['col']]` gives dataframe
- `sort_values(ascending=True)`

### Forgot to mention that we also covered
- Error handing
- Missing values with `.isnull()`
    - `.sum()` and `.sum(axis=1)`
    - There's also `.mean()` and `.std()` functions like this, e.g.

In [ ]:
import pandas as pd
amazonbooks = pd.read_csv('amazonbooks.csv', encoding="ISO-8859-1")

In [ ]:
amazonbooks

In [ ]:
amazonbooks.loc[ (amazonbooks.Hard_or_Paper=='P')&(amazonbooks.NumPages<100) ,('Title','Author','List Price')]

In [ ]:
amazonbooks.loc[ (amazonbooks.Hard_or_Paper=='P')&((amazonbooks.NumPages<50)|(amazonbooks.NumPages>600)),('Title','Author','List Price','NumPages')]


In [ ]:
amazonbooks.loc[ (amazonbooks.Hard_or_Paper!='H')&((amazonbooks.NumPages<50)|(amazonbooks.NumPages>600)),('Title','Author','List Price','NumPages')]


### Compound Conditon Boolean Selection

- `&`, `|`, `==`, `!=`

In [ ]:
# This would have been a good example to include as well, but I didn't think to do so
amazonbooks.loc[(amazonbooks.Hard_or_Paper=='P')&(amazonbooks.NumPages<100),
                ('Title','Author','List Price')].sort_values('List Price')


Also forgot to mention/summarize
- `.value_counts()`
- errata: ~~`.count()`~~ `.size()` in the first summary above